File: trees/_112_Path_Sum.java
public class _112_Path_Sum {

    //    Approach 1 - mistake is it stopped before leafNode
    class Solution {
        public boolean hasPathSum(TreeNode root, int targetSum) {
            if (root == null)
                return false;

            return helper(root, 0, targetSum);

        }

        public boolean helper(TreeNode node, int current, int targetSum) {
            if (node == null)
                return false;

            current += node.val;

            if (current == targetSum)
                return true;
            else if (current > targetSum)
                return false;
            else
                return helper(node.left, current, targetSum) || helper(node.right, current, targetSum);

        }
    }

    //    Approach 2 - all test cases passed
    class Solution2 {
        public boolean hasPathSum(TreeNode root, int targetSum) {
            if (root == null)
                return false;

            return helper(root, 0, targetSum);

        }

        public boolean helper(TreeNode node, int current, int targetSum) {
            if (node == null)
                return false;

            current += node.val;
            if (node.left == null && node.right == null) {
                if (current == targetSum)
                    return true;
            }

            return helper(node.left, current, targetSum) || helper(node.right, current, targetSum);

        }
    }
}

File: trees/_112_path_sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return helper(root, 0, targetSum);
    }

    bool helper(TreeNode* node, int sum, int targetSum) {
        if (node == nullptr)
            return false;
        sum += node->val;

        // if both nodes are null
        if (node->left == nullptr && node->right == nullptr) {
            return sum == targetSum;
        }
        return helper(node->left, sum, targetSum) ||
               helper(node->right, sum, targetSum);
    }
};